File: Python/163. Missing Ranges.py
class Solution:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        ans = list()
        if len(nums) == 0:
            if upper == lower:
                ans.append(str(lower))
            else:
                ans.append(str(lower)+"->"+str(upper))
        for i, n in enumerate(nums):
            if i == 0:
                if n - lower > 1:
                    ans.append(str(lower)+"->"+str(n-1))
                elif n - lower == 1:
                    ans.append(str(lower))
            else:
                if n - nums[i - 1] > 2:
                    ans.append(str(nums[i-1]+1)+"->"+str(n-1))
                elif n - nums[i-1] == 2:
                    ans.append(str(nums[i-1]+1))
            if i == len(nums) - 1:
                if upper - n > 1:
                    ans.append(str(n+1)+"->"+str(upper))
                elif upper - n == 1:
                    ans.append(str(upper))
        return ans

File: 2024/163. Missing Ranges.cpp
class Solution {
public:
    vector<vector<int>> findMissingRanges(vector<int>& nums, int lower, int upper) {
        if (nums.empty()) {
            return {{lower, upper}};
        }
        vector<vector<int>> res;
        if (nums[0] > lower) {
            res.push_back({lower, nums[0] - 1});
        }
        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] - nums[i - 1] != 1) {
                res.push_back({nums[i - 1] + 1, nums[i] - 1});
            }
        }
        if (nums.back() < upper) {
            res.push_back({nums.back() + 1, upper});
        }
        return res;
    }
};